kafka/kafka.go

```diff
@@ -242,34 +242,47 @@ func (t *subTopic) Subscribe(handler bps.Handler, options ...bps.SubOption) (bps
 	}
 
 	// synchronize handler access:
-	handler = &bps.SyncHandler{Handler: handler}
+	handler = bps.SafeHandler(handler)
 
-	// init closeable subscription/thread manager:
+	// init closeable subscription/thread group
 	sub := concurrent.NewGroup(context.Background())
 
 	// spawn partition-consuming threads:
 	for _, partition := range partitions {
-		pc, err := t.consumer.ConsumePartition(t.name, partition, initialOffset)
-		if err != nil {
+		if err := t.partition(sub, partition, initialOffset, handler, opts); err != nil {
 			_ = sub.Close()
 			return nil, fmt.Errorf("consume %s/%d partition: %w", t.name, partition, err)
 		}
+	}
 
-		sub.Go(func() {
-			defer pc.Close()
-
-			for {
-				select {
-				case <-sub.Done():
-					return
+	return sub, nil
+}
 
-				// TODO: what if one partition errors and it's Messages() chan is closed? Should we (try to) restart it?
-				case msg := <-pc.Messages():
```

Comment:
Basically, when `pc.Messages()` channel is closed, the last ms will be `nil` and `nil.Value` will panic. When consuming channels through `<-`, always do `msg, more := <-channel` instead of just `msg := <-channel`